staking/contracts/RariMineV3.sol

```diff
@@ -0,0 +1,149 @@
+    uint256 constant CLAIM_CLIFF_WEEKS     = 42;        // the meaning of life, the universe, and everything
+    uint256 constant CLAIM_SLOPE_WEEKS     = 42;        // the meaning of life, the universe, and everything
+    uint8   public constant VERSION               = 1;
+
+    mapping(address => uint) public claimed;
+
+    function __RariMineV3_init(IERC20Upgradeable _token, address _tokenOwner, IStaking _staking) external initializer {
+        __RariMineV3_init_unchained(_token, _tokenOwner, _staking);
+        __Ownable_init_unchained();
+        __Context_init_unchained();
+    }
+
+    function __RariMineV3_init_unchained(IERC20Upgradeable _token, address _tokenOwner, IStaking _staking) internal initializer {
+        token = _token;
+        tokenOwner = _tokenOwner;
+        staking = _staking;
+        _decimals = 18;
+        _name = "Rari Mine Claimed Tokens";
+        _symbol = "RariMineCT";
+    }
+
+    function claim(Balance memory _balance, uint8 v, bytes32 r, bytes32 s) public {
+        require(prepareMessage(_balance, address(this)).recover(v, r, s) == owner(), "owner should sign balances");
+
+        address recipient = _balance.recipient;
+        if (_msgSender() == recipient) {
+            uint toClaim = _balance.value.sub(claimed[recipient]);
+            require(toClaim > 0, "nothing to claim");
+            claimed[recipient] += _balance.value;
+            _totalSupply += _balance.value;
+
+            // claim rari tokens
+            uint claimAmount = toClaim.mul(CLAIM_FORMULA_CLAIM).div(CLAIM_FORMULA_DIVIDER);
+            require(token.transferFrom(tokenOwner, _msgSender(), claimAmount), "transfer to msg sender is not successful");
+            emit Claim(recipient, claimAmount);
+            emit Value(recipient, _balance.value);
+
+            // stake some tokens
+            uint stakeAmount = toClaim.sub(claimAmount);
+            uint slope = LibStakingMath.divUp(stakeAmount, CLAIM_SLOPE_WEEKS);
+            require(token.transferFrom(tokenOwner, address(this), stakeAmount), "transfer to RariMine is not successful");
+            require(token.approve(address(staking), stakeAmount), "approve is not successful");
+            staking.stake(_msgSender(), _msgSender(), stakeAmount, slope, CLAIM_CLIFF_WEEKS);
+            return;
+        }
+        
+        revert("_msgSender() is not the receipient");
+    }
+
+    function doOverride(Balance[] memory _balances) public onlyOwner {
+        for (uint i = 0; i < _balances.length; i++) {
+            claimed[_balances[i].recipient] = _balances[i].value;
+            emit Value(_balances[i].recipient, _balances[i].value);
+        }
+    }
+
+    function prepareMessage(Balance memory _balance, address _address) internal pure returns (string memory) {
+        uint256 id;
+        assembly {
+            id := chainid()
+        }
+        return toString(keccak256(abi.encode(_balance, _address, id, VERSION)));
```

Comment:
We need to have this in js code (prepareMessage)

Reply:
done
have a js function
tested with the tests